### PR TITLE
Add QThread/Python threading adaptor

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Adds a widget that adapts between Python threading and QThreads. This
provides a number of advantages described in the docstring for
associated classes.

The short-version is it gives us (much) better stack traces, allows us
to get rid of some bare catch exceptions and avoids users having to wire
up Qt signals and slots to simply run something off the GUI thread.

Whilst it's not present in this commit, this has been tested with the
ISIS SANS interface to verify the functionality with Qt and also fix up
some API usability problems noticed whilst implementing this

**To test:**
- Ensure unit tests pass
- Further testing will be present in the follow up PR which replaces the ISIS SANS QThread implementation with this widget

Part of #27895 

This does not require release notes because it's an internal change, though if when we convert users of QThread to this we should explicitly add it as Python stack traces improve greatly

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqt/utils/async_qt_adaptor.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        super(AsyncTaskQtAdaptor, self).__init__(
+            target=target, args=args, kwargs=kwargs,
+            success_cb=success_cb, error_cb=error_cb, finished_cb=finished_cb
+        )
+
+    @staticmethod
+    def _wrap_signal(signal: pyqtSignal, slot: Optional[Callable]) -> Optional[Callable]:
+        """
+        Wraps a given signal so that the thread calls emit rather than the target directly.
+        This allows us to complete the slot on the target, rather than the
```
Comment (clayton-tom) [inline qt/python/mantidqt/utils/async_qt_adaptor.py:170]: Think you might've forgotten to finish this sentence

**Review by clayton-tom — CHANGES_REQUESTED** [review]
New unit looks great and the tests pass, just one comment on the docstring

File: qt/python/mantidqt/utils/async_qt_adaptor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DavidFair) [reply qt/python/mantidqt/utils/async_qt_adaptor.py]: Good spot, I had got distracted looking into a test instead :+1:

**Review by clayton-tom — DISMISSED** [review]
All looks good :+1:

File: qt/python/mantidqt/utils/async_qt_adaptor.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+from functools import wraps
+from typing import Any, Optional, Callable
+
+from PyQt5 import QtCore
```
Comment (martyngigg) [inline qt/python/mantidqt/utils/async_qt_adaptor.py:10]: Can we swap these for `qtpy` imports?

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — DISMISSED** [review]

File: qt/python/mantidqt/utils/async_qt_adaptor.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+from functools import wraps
+from typing import Any, Optional, Callable
+
+from qtpy import QtCore
+from qtpy.QtCore import QObject, pyqtSignal
```
Comment (martyngigg) [inline qt/python/mantidqt/utils/async_qt_adaptor.py:11]: I forget everytime that in `qtpy` the object is called `Signal` and not `pyqtSignal` 🤦🏻 
 ```suggestion
from qtpy.QtCore import QObject, Signal
```

Comment (DavidFair) [reply qt/python/mantidqt/utils/async_qt_adaptor.py]: Weirdly this worked for me locally as I tested before pushing :confused:

**Review by martyngigg — APPROVED** [review]

**Pull request merged**